crates/sim2h/src/lib.rs

```diff
@@ -260,49 +231,565 @@ impl Sim2hHandle {
+            WireMessage::Status => return spawn_handle_message_status(sim2h_handle, uri, signer),
+            WireMessage::Hello(version) => {
+                return spawn_handle_message_hello(sim2h_handle, uri, signer, version)
+            }
+            WireMessage::ClientToLib3h(ClientToLib3h::JoinSpace(data)) => {
+                return spawn_handle_message_join_space(sim2h_handle, uri, signer, data)
+            }
+            message @ _ => message,
+        };
+
+        // you have to be in a space to proceed further
+
+        tokio::task::spawn(async move {
+            // -- right now each agent can only be part of a single space :/ --
+
+            let state = sim2h_handle.im_state().get_clone().await;
+            let (agent_id, space_hash) = match state.get_space_info_from_uri(&uri) {
+                None => {
+                    error!("uri has not joined space, cannoct proceed {}", uri);
+                    return;
+                }
+                Some((agent_id, space_hash)) => (agent_id, space_hash),
+            };
+
+            if *agent_id != signer {
+                error!(
+                    "signer {} does not match joined agent {:?}",
+                    signer, agent_id
+                );
+                return;
+            }
+
+            match message {
+                WireMessage::ClientToLib3h(ClientToLib3h::LeaveSpace(_data)) => {
+                    // for now, just disconnect on LeaveSpace
+                    sim2h_handle.disconnect(vec![uri.clone()]);
+                    return;
+                }
+                WireMessage::ClientToLib3h(ClientToLib3h::SendDirectMessage(dm_data)) => {
+                    return spawn_handle_message_send_dm(
+                        sim2h_handle,
+                        uri,
+                        signer,
+                        space_hash,
+                        dm_data,
+                    );
+                }
+                WireMessage::Lib3hToClientResponse(
+                    Lib3hToClientResponse::HandleSendDirectMessageResult(dm_data),
+                ) => {
+                    return spawn_handle_message_send_dm_result(
+                        sim2h_handle,
+                        uri,
+                        signer,
+                        space_hash,
+                        dm_data,
+                    );
+                }
+                WireMessage::ClientToLib3h(ClientToLib3h::PublishEntry(data)) => {
+                    return spawn_handle_message_publish_entry(
+                        sim2h_handle,
+                        uri,
+                        signer,
+                        space_hash,
+                        data,
+                    );
+                }
+                WireMessage::Lib3hToClientResponse(
+                    Lib3hToClientResponse::HandleGetAuthoringEntryListResult(list_data),
+                ) => {
+                    return spawn_handle_message_list_data(
```

Comment:
this isn't sufficient - depending on how the sharding works out, gossip won't know to go here for fetches - despite the inefficiency we should go ahead and send fetch requests for all these entries and they can be ignore on receipt if everybody already has them.

Reply:
fixed